### ISSUE TITLE
Using wxBitmapBundle with wxTreeCtrl doesn't work in some cases

### ISSUE THREAD

**Issue by craftyjon** [wxWidgets/wxWidgets#23994]
### Description

When attempting to move a wxTreeCtrl from legacy images to wxBitmapBundle, I hit an assertion with the following stack trace:

```
<...>
wxBitmap::GetSubBitmap(const wxRect &) const bitmap.cpp:968
wxGenericImageList::GetImageListBitmap(const wxBitmap &) const imaglist.cpp:94
wxGenericImageList::Add(const wxBitmap &) imaglist.cpp:135
wxBitmapBundle::CreateImageList(wxWindow *, const wxVector<…> &) bmpbndl.cpp:688
[Inlined] wxWithImages::GetUpdatedImageListFor(wxWindow *) withimages.h:123
wxGenericTreeCtrl::OnImagesChanged() treectlg.cpp:2482
wxWithImages::SetImages(const wxVector<…> &) withimages.h:79
<application code removed>
```

The assertion is firing because `GetSubBitmap` is being called with a rect of 16x16, but the bitmap in question at this point in the call stack is only 11x11 for some reason.

Looking higher up the call stack, it seems to me like the wxBitmapBundle conversion may be conflicting with some code to handle wxBitmaps with "more than one bitmap".

The wxBitmapBundle I created in this case has a base resolution of 24x24 pixels, so in `wxGenericImageList::Add`, `bitmapSize` is 24x24.  This triggers the "more than one bitmap" code, which then resizes it to 17x17, and then 11x11 once more down the call stack.

I'm not quite sure what the ideal behavior here is as I don't understand the "more than one bitmap" code's intended functionality, but it seems possible to me that this should be entirely skipped in the case that the images are coming from a wxBitmapBundle, and instead the bundle should be tasked with returning an icon at the requested size from the treectrl (in this case, a 16x16 icon is actually present in the bundle, it's just not ever used because the "primary" size of the bundle is 24x24 for other reasons, and the `wxBitmapBundle` APIs aren't used to retrieve a scaled image).

Expectation:  Any `wxBitmapBundle` for which `wxBitmapBundle::IsOk()` returns true should be able to be passed in to `wxTreeCtrl::SetImages` and rendered corrrectly with no assertions.

### Platform and version information

- wxWidgets version  you use: 3.2.3
- wxWidgets port  you use: wxOSX
- OS  and its version: macOS 13.5.1

**Comment by vadz** [comment]
Sorry, I'm confused about the different sizes involved here. The simplest way to clear this confusion would be to give me some simple way to reproduce the bug, but otherwise I'd like to know:

1. What exactly is the bundle contents? Just a single 24x24 bitmap or something else?
2. What is your DPI scaling? 150%, right?
3. What is the size of the bitmap passed to `wxImageList::Add()`? It looks like it ought to be 24x24 physical but then the logical `bitmapSize` should be 16x16 at 150% scaling, but you say it's 24x24 itself, which means that I misunderstand something here...


_Historical Note_: The "more than one bitmap" is about using the standard Windows trick (technique?) to add just one very wide bitmap containing all the application icons to the image list (this used to be a worthy optimization back in 16-bit Windows days when GDI had a very low bitmaps limit and lost its purpose decades ago, but was cargo culted in so much code that I'm sure people still continue doing this today without having any idea why).

**Comment by craftyjon** [comment]
> The simplest way to clear this confusion would be to give me some simple way to reproduce the bug

I will work on this once I have time.

> What exactly is the bundle contents? Just a single 24x24 bitmap or something else?

Multiple images, but the 24x24 is the first one added, which wxBitmapBundle seems to treat specially.

> What is your DPI scaling? 150%, right?

200% on this mac, however I'm not sure this issue is dependent on DPI scaling

> What is the size of the bitmap passed to wxImageList::Add()

Let me say a bit more about what I think is going on: I think 16x16 (logical pixels) is the "desired size" of the icon from the tree control, which I assume is somewhat arbitrary but relating to the size of each list entry in the tree.  The desired effect here is to pass in a list of bitmap bundles, and have the tree control request the 16x16 logical, which will return a 32x32 bitmap on my system but that is besides the point.  The 24x24 does not come from 150% DPI scaling, it comes from the fact that the 24x24 input image is the first one added to the bitmap bundle, and somewhere in this stack, the first image size is used rather than the requested size coming from the caller (the tree list)

**Comment by craftyjon** [comment]
> Just a single 24x24 bitmap or something else?

I was wrong about this, I did further debugging and found that due to an error on my side, some bundles contained a single 24x24 image, and these seem to be the ones causing the assert.

So:

`wxBitmapBundle::CreateImageList`: `size` is 16x16 ("consensus" size I guess?)
`GetBitmap` must have been called on the bundle containing only a 24x24 bitmap with size == 16x16
`wxGenericImageList::Add`: `bitmap.GetLogicalSize()` returns 24x24

I presume this is because of:
```
    // Ensure that the returned bitmap uses the scale factor such that it takes
    // the same space, in logical pixels, as the bitmap in the default size.
```

**Comment by vadz** [comment]
The consensus size should really be 24x24 in logical units if there are bundles using 16x16 and 24x24, see the (passing) test here:

https://github.com/wxWidgets/wxWidgets/blob/dc6a0c069bdde714f22cb0459efd9ad186ce7179/tests/graphics/bmpbundle.cpp#L493-L496

And I've tried adding

```diff
diff --git a/tests/graphics/bmpbundle.cpp b/tests/graphics/bmpbundle.cpp
index bc890fb9e2..dac7fb96f2 100644
--- a/tests/graphics/bmpbundle.cpp
+++ b/tests/graphics/bmpbundle.cpp
@@ -14,8 +14,10 @@
 
 #include "wx/bmpbndl.h"
 
+#include "wx/app.h"
 #include "wx/artprov.h"
 #include "wx/dcmemory.h"
+#include "wx/imaglist.h"
 
 #include "asserthelper.h"
 
@@ -503,6 +505,19 @@ TEST_CASE("BitmapBundle::GetConsensusSize", "[bmpbundle]")
     CHECK( bundles.GetConsensusSize(1.5) == 16 );
 }
 
+TEST_CASE("BitmapBundle::ImageList", "[bmpbundle][imagelist]")
+{
+    wxVector<wxBitmapBundle> images;
+    images.push_back(wxBitmapBundle::FromBitmap(wxBitmap(16, 16)));
+    images.push_back(wxBitmapBundle::FromBitmap(wxBitmap(24, 24)));
+
+    std::unique_ptr<wxImageList>
+        iml(wxBitmapBundle::CreateImageList(wxTheApp->GetTopWindow(), images));
+    REQUIRE( iml );
+
+    CHECK( iml->GetSize() == wxSize(24, 24) );
+}
+
 // This test is not really related to wxBitmapBundle, but is just here because
 // this file already uses wxArtProvider and we don't have any tests for it
 // specifically right now.
```

but it passes too, so there must be something else/more to it than this.

**Comment by craftyjon** [comment]
I'll keep trying to debug this locally

For reference, on the 200% scaling Mac, here's what it looks like at present.  The state images are 16x16 at the moment and they are fine.  The top non-state icon is in a bundle that contains 24x24 and 16x16 (as well as other sizes).  The bottom two icons are in bundles that only contain 24x24.  Having this state (in KiCad at least) results in asserts, as well as this mismatch in actual rendered icon size, and blurriness in the "24x24" images (maybe they are actually the 16x16 images scaled up to 24x24?  will need to investigate)

<img width="82" alt="image" src="https://github.com/wxWidgets/wxWidgets/assets/116737/8b412110-327b-4b4b-82c1-53c759c1b212">

**Comment by craftyjon** [comment]
I think this may indeed be dependent on running on a HiDPI system.

If I have a set of bundles where:

- some of them have two bitmaps: 16x16, 32x32
- others of them have one bitmap: 24x24
- there are more of the first kind than the second kind

and then I run through this tree control setup on my system (scale == 2)

`GetPreferredBitmapSizeAtScale` will return 32x32 for the first case, but 24x24 for the second case
Because there are more of the first case, the consensus size will be 32x32, which is divided by the scale in CreateImageList to get a target size of 16x16.  So far, this all seems probably fine.

Now: `wxBitmapBundle::GetBitmap( {16, 16} )` will be called on the bundles that only have the 24x24.  This will return a bitmap with its scale factor set to 2/3.  This seems wrong, because the assertion in the comment here seems inaccurate for this use case: we don't actually want the bitmap to take the same space in logical pixels as the default size (24x24), we want it to take the same space in logical pixels as the consensus size of the group (16x16).  Note that this assumption baked in to GetBitmap has also caused us to have to work around this in other places in KiCad, where we are requesting a specific size and really don't want a different bitmap scale factor to be set here.

Now in `wxGenericImageList::Add` when we get to one of these 24x24 bitmaps, `bitmapSize` is 24x24, but `m_size` is 16x16.  Because of this, we enter the "sub bitmap" codepath and call `GetSubBitmap` with a rect of 16x16.

The bitmap returned by this codepath has a logical size of 17x17 now, probably due to a rounding issue in this unnecessary copying of the bitmap and handling the logical size / scaling factor.  This 17x17 bitmap is passed to `GetImageListBitmap`.  Since this is larger than 16x16, we call `GetSubBitmap` *again*, where the assert fires because now the logical size is only 11x11px and we're asking for a 16x16 rect.

I am getting a little lost with all the resizing, scale factors, etc so I haven't quite pieced together why the rescales are happening the way they are, and it seems like maybe there are multiple issues here.  But the main cause of the issue for me seems to be that `GetBitmap({16, 16})` will return a bitmap that thinks it is supposed to be 24x24 when the bundle contains only a 24x24 bitmap.  It seems like what is really needed is either an API in wxBitmapBundle that returns a bitmap without assuming that you wanted it rescaled to the default size, or for the call site to know that the bundle may do this, and reset the scaling factor to 1.

**Comment by vadz** [comment]
OK, thanks, I'll try to have a look at this soon. I think the key is indeed having both 16/32 and 24 bundles, so I'll test this. FWIW, there is no chance that things look decent in this setup as some images will get scaled by a factor of 1.5 which can but look ugly, so I think you will still need to harmonize the sizes you use. But wx still shouldn't assert, of course.

**Comment by craftyjon** [comment]
> FWIW, there is no chance that things look decent in this setup as some images will get scaled by a factor of 1.5 which can but look ugly, so I think you will still need to harmonize the sizes you use. But wx still shouldn't assert, of course.

Yes, the right "fix" for this on my side is to provide the 16 and 32 pixel versions of the 24-pixel icons of course.  But I think there are two problems on the wx side:

1) It shouldn't assert
2) Even when ignoring the assert, or avoiding it by patching out the scaling factor adjustment, the 24x24 bitmap is not properly scaled down to 16x16 for rendering (or rather, scaled up to 32x32 device pixels in my case).  So, the tree icons have non-uniform sizes, as the ones that contain 32x32 images are rendered in the target size (16x16 logical pixels) but the ones that contain only 24x24 are rendered larger (looks like 24x24 logical pixels)

**Comment by vadz** [comment]
Turns out that this is a Mac-specific bug, this test case:

```cpp
TEST_CASE("BitmapBundle::ImageList", "[bmpbundle][imagelist]")
{
    wxVector<wxBitmapBundle> images;
    images.push_back(wxBitmapBundle::FromBitmaps(wxBitmap(16, 16), wxBitmap(32, 32)));
    images.push_back(wxBitmapBundle::FromBitmap(wxBitmap(24, 24)));
    images.push_back(wxBitmapBundle::FromBitmaps(wxBitmap(16, 16), wxBitmap(32, 32)));

    // There are 2 bundles with preferred size of 32x32, so they should win.
    wxSize size = wxBitmapBundle::GetConsensusSizeFor(2.0, images);
    CHECK( size == wxSize(32, 32) );

    size /= 2.0;
    wxImageList iml(size.x, size.y);
    iml.Add(images[0].GetBitmap(size));
    iml.Add(images[1].GetBitmap(size));
}
```

passes with wxGTK, but fails with wxOSX. Debugging...

**Comment by vadz** [comment]
So you were right that there were multiple problems here, but there are even more of them than you thought. All of them are related, as usual, to the confusion between logical and physical sizes and I'd like to write them down here before starting fixing them to check that I'm not confused again myself:

1. The fundamental problem is that `wxImageList` is just incompatible with using DPI-dependent bitmaps and can't be changed to support them. Proof: it can be shared between multiple windows (this is its original raison d'être, in fact), but these windows can be on different displays using different DPI, so this just can't work. Yet there have been several commits trying to make it work in high DPI, notably eb52e86553 (Take into account scale factor of images added to wxImageList, 2021-04-03), b6d305e4f2 (Assure that all images in the generic wxImageList have the same scale factor, 2021-04-03) and ba588fd072 (Take into account scale actor while resizing the image in generic wxImageList, 2021-04-03). To make things even messier, some of them were reverted later, but arguably we should stop trying to support high DPI in this code completely because it has no chance of working without `wxImageList` having its own scale factor (which it can't, as explained just above). I'm not sure what to do here by now to be honest.
1. The immediate problem here is that `wxImageList` is still used by `wxGenericTreeCtrl` internally, even when `wxBitmapBundle`-based API is used. With #24000 there should be no need to use it any longer, so I'm going to remove all the remaining uses of it in this code and then `wxGenericTreeCtrl` won't be affected by the problem above, whatever we do (or don't do about it).
1. The difference in behaviour between wxOSX and wxGTK is due to using rounding in `wxBitmap::GetSubBitmap()` in the former and truncation in the latter, resulting in a subbitmap of size 16 of the bitmap with scale factor ⅔ being 11px in one and 10px in the other one which, in turn, resulted in it getting of size 15 or 17 (rounding is always used in this direction) after applying the scale factor to it again, which meant that it was greater than 16 only in wxOSX but not wxGTK. This can be fixed by making wxGTK use rounding too, as I think it's better than truncation even though I'm not 100% sure about it. But I'm sure it's better to have the same behaviour under all platforms, so one of them needs to change and I suggest changing wxGTK and will do it.
1. Final, and relatively minor, problem is that the assert in wxOSX version of `GetSubBitmap()` is wrong because it compares logical rectangle size passed to it with physical bitmap size. This is simple to fix, and wxGTK already does this correctly, and I'll do it too.

Please let me know if anybody sees anything wrong in what I wrote here or any suggestions about what to do with `wxImageList`.

**Comment by craftyjon** [comment]
To be honest it's not clear to me why `wxImageList` exists: why couldn't the tree and list controls just take a `wxVector<wxBitmapBundle>`?

**Comment by vadz** [comment]
> To be honest it's not clear to me why `wxImageList` exists

By now, only for backwards compatibility. Originally, because `wx{List,Tree}Ctrl` API was based on their Win32 versions (back in the nineties...) and they used `HIMAGELIST` which was used for resource conservation reasons that probably didn't make much sense even then.

**Comment by vadz** [comment]
I've addressed the point (2) above in the new commits added to #24000 and (3) and (4) are taken care of in #24010.

Point (1) still remains, but I don't think it can be satisfactorily solved anyhow, and it doesn't seem like a priority now that (2) is (or at least should be) solved.

Please let me know if you still see any problems with the latest #24000 (you don't need #24010, but merging it too shouldn't hurt neither).

**Comment by vadz** [comment]
I've finally created another PR to fix this. Unfortunately this one can't be backported to 3.2 because it's a clearly incompatible change, but you could still use it in your own branch, of course, and at least it will behave correctly even with the mismatched bitmaps in 3.3.
